src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -168,13 +169,89 @@ private void OnEventSourceChanged(object? _1, TaggerEventArgs _2)
+                // more new work comes in.
+                if (!frozenPartialSemantics)
+                    nonFrozenComputationToken = nextToken;
+            }
+
+            _eventChangeQueue.AddWork(
+                new TagSourceQueueItem(highPriority, frozenPartialSemantics, nonFrozenComputationToken),
+                _dataSource.CancelOnNewWork);
+        }
+
+        private async ValueTask ProcessEventChangeAsync(
+            ImmutableSegmentedList<TagSourceQueueItem> changes, CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfTrue(changes.IsEmpty);
+
             // If any of the requests was high priority, then compute at that speed.
-            var highPriority = changes.Contains(true);
-            return new ValueTask<VoidResult>(RecomputeTagsAsync(highPriority, cancellationToken));
+            var highPriority = changes.Contains(x => x.HighPriority);
+
+            // If any of the requests are for frozen partial, then we do compute with frozen partial semantics.  We
+            // always want these "fast but inaccurate" passes to happen first.  That pass will then enqueue the work
+            // to do the slow-but-accurate pass.
+            var frozenPartialSemantics = changes.Contains(t => t.FrozenPartialSemantics);
+
+            if (frozenPartialSemantics)
+            {
+                // If we were asking for frozen partial semantics, then just proceed as normal, getting those tags
+                // quickly, but inaccurately.
+                await RecomputeTagsAsync(highPriority, frozenPartialSemantics: true, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                if (!_dataSource.SupportsFrozenPartialSemantics)
+                {
+                    // We're asking for normal expensive full tags, and this tagger doesn't support frozen partial
+                    // tagging anyways, so just proceed as normal, asking for expensive tags.
+                    await RecomputeTagsAsync(highPriority, frozenPartialSemantics: false, cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    // We're asking for expensive tags, and this tagger supports frozen partial tags.  Kick off the work
+                    // to do this expensive tagging, but attach ourselves to the requested cancellation token so this
+                    // expensive work can be canceled if new requests for frozen partial work come in.
+
+                    // We must have at least one request asking for full semantics.
+                    Contract.ThrowIfFalse(changes.Any(c => !c.FrozenPartialSemantics));
+
+                    // All those requests should have cancellation tokens provided with them.
+                    Contract.ThrowIfFalse(changes.Where(c => !c.FrozenPartialSemantics).All(c => c.NonFrozenComputationToken != null));
```

Comment:
`Contract.ThrowIfFalse` is a runtime assertion (not conditioned on `DEBUG`). The `Where` on this line is redundant due to the previous check.

```suggestion
                    Contract.ThrowIfFalse(changes.All(c => c.NonFrozenComputationToken != null));
```

Reply:
done.